### PR TITLE
Sharded Plugin 3/n: Expose step input to DDP plugin

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Relates to #4178 

ShardedDDP needs to move the device to the GPU before passing through to the model. This is not the case for DDP which handles this internally. Allows other custom DDP parallel plugins an opportunity to modify the input before passing into the wrapper!

cc @williamFalcon @tchaton @awaelchli @Borda 

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/ddp2_accelerator.py [hunk 1]
```diff
@@ -61,21 +61,23 @@ def train(self):
         return self.ddp_train(process_idx=self.task_idx, mp_queue=None, model=model)
 
     def training_step(self, args):
+        return self._step(args)
```
Comment (williamFalcon) [inline pytorch_lightning/accelerators/ddp2_accelerator.py:64]: nice. this is better haha.

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 2]
```diff
@@ -62,3 +62,14 @@ def configure_ddp(self, model, device_ids):
             **self._ddp_kwargs,
         )
         return model
+
+    def input_to_device(self, args: Any, model: LightningModule):
```
Comment (williamFalcon) [inline pytorch_lightning/plugins/ddp_plugin.py:66]: can you give an example of this? i'm not understanding this method.

Comment (williamFalcon) [reply pytorch_lightning/plugins/ddp_plugin.py:66]: also, if this call is time dependent, can this be made into a hook?

@ananthsub brought up the point that all accelerators should have the standard hooks as well. that should also be allocated to the next 1 or 2 sprints @edenafek.

maybe `on_before_forward`?

Comment (SeanNaren) [reply pytorch_lightning/plugins/ddp_plugin.py:66]: yeah that's a neat idea, let me make it a hook for now within the plugin with the name `on_before_forward`! This is what I do in the sharded plugin:

https://github.com/PyTorchLightning/pytorch-lightning/pull/4178/files#diff-49267e3a168db4eb60775a18da157240d63c25cab73eb45e732a7d45db7f0fa8R50

Basically this call is skipped for DDP because DDP handles device movement internally, sharded DDP needs device to be moved up front

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/accelerators/ddp2_accelerator.py [hunk 3]
```diff
@@ -61,21 +61,23 @@ def train(self):
...
+
+    def validation_step(self, args):
+        return self._step(args)
```
Comment (Borda) [inline pytorch_lightning/accelerators/ddp2_accelerator.py:67]:
```suggestion
    def validation_step(self, *args):
        return self._step(*args)
```
or can we define what is the type? and how about `**kwargs`?

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 4]
```diff
@@ -62,3 +62,14 @@ def configure_ddp(self, model, device_ids):
...
         return model
+
+    def on_before_forward(self, args: Any, model: LightningModule):
+        """
+        Override to handle custom input to device logic. For DDP, no logic is required as this is handled internally
+        within the DDP wrapper.
+        Args:
+            args: Inputs to the model.
+            model: Model to train.
+        Returns: args moved to correct device if needed.
```
Comment (Borda) [inline pytorch_lightning/plugins/ddp_plugin.py:73]:
```suggestion
        
        Args:
            args: Inputs to the model.
            model: Model to train.
        
        Returns: args moved to correct device if needed.
```
can we add an example of such inputs?

File: pytorch_lightning/accelerators/ddp2_accelerator.py (hunk 3 above)
Comment (SeanNaren) [reply pytorch_lightning/accelerators/ddp2_accelerator.py:67]: I think I'll leave this for a separate PR as I don't want to mess with the accelerator logic too much... kwargs are not explicitly passed into the training step upstream, so this needs to be an involved change

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/accelerators/ddp_accelerator.py [hunk 5]
```diff
@@ -151,21 +151,23 @@ def train(self):
...
+        return self._step(args)
+
+    def validation_step(self, args):
+        return self._step(args)
+
+    def test_step(self, args):
+        return self._step(args)
+
+    def _step(self, args):
+        args = self.ddp_plugin.on_before_forward(self.trainer.get_model(), *args)
```
Comment (williamFalcon) [inline pytorch_lightning/accelerators/ddp_accelerator.py:163]: love it

**Review by williamFalcon — APPROVED** [review]
